### ISSUE TITLE
SSL Termination on AWS NLB to avoid managing certs in contour

### ISSUE THREAD

**Issue by farshad-hobsons** [projectcontour/contour#2441]
AWS came out with [TLS Termination for Network Load Balancers](https://aws.amazon.com/blogs/aws/new-tls-termination-for-network-load-balancers/) on Jan 24, 2019. In order to avoid managing ssl certs manually or through cert-manager, I would prefet to use ssl termination on aws-nlb while leveraging the other nice contour features. So basically, I am looking whether this sort of architecture is possible:
```
https-traffic -> aws-nlb -> http-traffic-to-envoy
```
I have successfully tested this same type of functionality in nginx-controller using the following annotations with aws-nlb:
```
    service.beta.kubernetes.io/aws-load-balancer-backend-protocol: "ssl"
    service.beta.kubernetes.io/aws-load-balancer-ssl-ports: "443"
    service.beta.kubernetes.io/aws-load-balancer-type: nlb
    service.beta.kubernetes.io/aws-load-balancer-ssl-cert: "mycertARN"
```

Unfortunately, following the [deploy-aws-nlb.md](https://github.com/projectcontour/contour/blob/master/site/_guides/deploy-aws-nlb.md) with the above annotations did not work (traffic does not make it to envoy) for contour. I also tried testing by  setting: `hostNetwork: true` and `dnsPolicy: ClusterFirstWithHostNet` on my Envoy pod definition as per [contour's Host Networking docs](https://projectcontour.io/docs/master/deploy-options/)

If the above architecture is possible, it would be help the community if we can update the aws nlb docs. If its not possible, it would be good to know the technical reasons and if this would be possible in future contour releases.

**Comment by stevesloka** [comment]
Hey @farshad-hobsons I took some time this morning to setup your scenario. I did get it working, the TLDR is that I changed the TLS Listener on port 443 for the NLB to the same that used on port 80. 

The reason for this change is that Contour only sets up a TLS listener on Envoy when a TLS secret exists. Since there's no secret (i.e. you're not terminating TLS at Envoy), the default group in AWS sends traffic to the port 443 nodePort which isn't listening. 

Here's a sample of my setup:
![image](https://user-images.githubusercontent.com/1048184/81417867-d6ab9380-9119-11ea-99c2-2513bed58522.png)

This would also be great to add to the docs on how to configure this properly from the start. I'll pop open a new issue for that.

**Comment by farshad-hobsons** [comment]
thanks @stevesloka for providing this workaround.

**Comment by psehgal** [comment]
This is a great workaround. But is there any way to manage this through the manifests for contour?

**Comment by WKuipers** [comment]
I've run into this issue too while deploying through Helm. If that chart would allow customizing targetPort for https on the envoy service it would be awesome.

**Comment by psehgal** [comment]
@WKuipers I had no idea there was a helm chart for install this. Thank you for sharing. It looks like this value might actually be driven by a variable you can configure in your `values.yaml`

```
  ports:
    - name: http
      port: {{ .Values.envoy.service.ports.http }}
      protocol: TCP
      targetPort: http
      {{- if and (or (eq .Values.envoy.service.type "NodePort") (eq .Values.envoy.service.type "LoadBalancer")) (not (empty .Values.envoy.service.nodePorts.http)) }}
      nodePort: {{ .Values.envoy.service.nodePorts.http }}
      {{- else if eq .Values.envoy.service.type "ClusterIP" }}
      nodePort: null
      {{- end }}
    - name: https
      port: {{ .Values.envoy.service.ports.https }}
      protocol: TCP
      targetPort: https
      {{- if and (or (eq .Values.envoy.service.type "NodePort") (eq .Values.envoy.service.type "LoadBalancer")) (not (empty .Values.envoy.service.nodePorts.https)) }}
      nodePort: {{ .Values.envoy.service.nodePorts.https }}
      {{- else if eq .Values.envoy.service.type "ClusterIP" }}
      nodePort: null
      {{- end }}
```

**Comment by youngnick** [comment]
I see this request, maybe @georgegoh could comment on how to log an issue for the Helm chart?

**Comment by ectogon** [comment]
Has anyone tried this in combination with `nginx.ingress.kubernetes.io/force-ssl-redirect: "true"`?

**Comment by abhinavk07** [comment]
```
apiVersion: v1
kind: Service
metadata:
  name: elm-nlb
  namespace: demo
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-internal: "true"
    service.beta.kubernetes.io/aws-load-balancer-type: nlb
    service.beta.kubernetes.io/aws-load-balancer-ssl-cert: ***
    service.beta.kubernetes.io/aws-load-balancer-backend-protocol: ssl
    service.beta.kubernetes.io/aws-load-balancer-ssl-ports: "443"
    service.beta.kubernetes.io/aws-load-balancer-cross-zone-load-balancing-enabled: 'true'
spec:
  ports:
    - name: http
      port: 80
      targetPort: 80
      protocol: TCP
    - name: https
      port: 443
      targetPort: 80
      protocol: TCP
  type: LoadBalancer
  selector:
    app: demo-elm
```

Above config for NLB isn't working for me. I'm able to reach the application but the SSL cert doesn't work. I'm using kubernetes 1.14, can anyone suggest a workaround?

**Comment by abhinavk07** [comment]
```
apiVersion: v1
kind: Service
metadata:
  name: **
  namespace: demo
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-internal: "true"
    service.beta.kubernetes.io/aws-load-balancer-type: nlb
    service.beta.kubernetes.io/aws-load-balancer-ssl-cert: *****
    service.beta.kubernetes.io/aws-load-balancer-backend-protocol: ssl
    service.beta.kubernetes.io/aws-load-balancer-ssl-ports: "443"
    service.beta.kubernetes.io/aws-load-balancer-cross-zone-load-balancing-enabled: 'true'
spec:
  ports:
    - name: http
      port: 80
      targetPort: 80
      protocol: TCP
    - name: https
      port: 443
      targetPort: 80
      protocol: TCP
  type: LoadBalancer
  selector:
    app: **
```

Above config for NLB isn't working for me. I'm able to reach the application but the SSL cert doesn't work. I'm using kubernetes 1.14, can anyone suggest a workaround?

**Comment by youngnick** [comment]
Hi @abhinavk07, could you log a separate issue for that one? As well as the Kubernetes version, could you put in the Contour and Envoy versions you're using? I don't know much about these annotations, but we can try and give you a hand once we have that.

**Comment by rdgacarvalho** [comment]
Hi,
I am facing the same issue even after followed [deploy-aws-tls-nlb](https://projectcontour.io/guides/deploy-aws-tls-nlb/) guidance.
___

My environment

> AWS EKS - v1.20.7-eks-135321

> Helm chart - Contour-5.0.1 
```
      envoy:
        hostNetwork: true
        dnsPolicy: ClusterFirstWithHostNet
```
___

Kubernetes:  Service description
```
Name:                     contour-ingress-envoy
Namespace:                ingress
Labels:                   app.kubernetes.io/component=envoy
                          app.kubernetes.io/instance=contour-ingress
                          app.kubernetes.io/managed-by=Helm
                          app.kubernetes.io/name=contour
                          helm.sh/chart=contour-5.0.1
Annotations:              meta.helm.sh/release-name: contour-ingress
                          meta.helm.sh/release-namespace: ingress
                          service.beta.kubernetes.io/aws-load-balancer-additional-resource-tags: Environment=dev,Maintainers=Team:DevOps
                          service.beta.kubernetes.io/aws-load-balancer-connection-idle-timeout: 3600
                          service.beta.kubernetes.io/aws-load-balancer-ssl-cert: arn:xxxx-xxxx-xxx-xxx
                          service.beta.kubernetes.io/aws-load-balancer-ssl-ports: 443
                          service.beta.kubernetes.io/aws-load-balancer-type: nlb
Selector:                 app.kubernetes.io/component=envoy,app.kubernetes.io/instance=contour-ingress,app.kubernetes.io/name=contour
Type:                     LoadBalancer
```
___

Testing HTTP(s)/TLS via **curl**

HTTP - 200 OK
```
* TCP_NODELAY set
* Connected to whoami-contour.xxxx.xxx.xxxxx (x.x.x.x) port 80 (#0)
> HEAD / HTTP/1.1
> Host: whoami-contour.xxxx.xxx.xxxxx 
> User-Agent: curl/7.64.1
> Accept: */*
>
< HTTP/1.1 200 OK
HTTP/1.1 200 OK
< date: Thu, 09 Sep 2021 14:46:43 GMT
date: Thu, 09 Sep 2021 14:46:43 GMT
< content-length: 450
content-length: 450
< content-type: text/plain; charset=utf-8
content-type: text/plain; charset=utf-8
< x-envoy-upstream-service-time: 0
x-envoy-upstream-service-time: 0
< vary: Accept-Encoding
vary: Accept-Encoding
< server: envoy
server: envoy
```
Contour log:
```
[2021-09-09T14:46:29.143Z] "HEAD / HTTP/1.1" 200 - 0 0 1 1 "48.78.116.4" "curl/7.64.1" "8de496be-e584-40a1-8a09-4e2f170cb061"
```

HTTPS - timeout
```
```

Any suggestion or workaround is welcome? 

Thank you

**Comment by youngnick** [comment]
Hi @rdgacarvalho, could you please log a separate issue for this? It's easier to keep discussion tidy when we do that. From a quick look at what you've sent, it seems like everything is working, but I'd need to check over the guide myself.

**Comment by rdgacarvalho** [comment]
@youngnick I have opened the following PR https://github.com/bitnami/charts/pull/7571 which should allow us to use the workaround suggested by @psehgal much easier than managing the service template directly.

**Comment by youngnick** [comment]
That's a much better solution, thanks!

**Comment by gregkoganvmm** [comment]
This does not seem to work 100% (1.19.1, bitnami chart version 7.2.0). Setting the port to 80 (as indicated on in the bitnami chart) seems to break both http/https (possibly due to security of the container due to running as a non-root user), setting both ports to 8080 works better (can hit both ports successfully), but there are a couple of issues. FYI, I did not fully follow the guide because aws-load-balancer-controller conveniently does not yet work with kuberentes 1.22 (until version 2.4.0 allegedly) though I am not sure that would make a difference since we can make the changes manually.

1. Once proxy protocol is enabled (as the guide is recommending) on both target groups, it appears that envoy is not passing along the proxy headers (e.g. `X-Forwarded-Proto`), so if an application serves both http and https traffic and relies on the headers to determine this, it does not work - I always get http urls in the html response.
2. `ingress.kubernetes.io/force-ssl-redirect` results in an infinite loop likely due to issue 1

I went through guides/documentation on the Contour site and not seeing anything that can help here... I would really like to stop using AWS ELBs and nginx ingress and switch to Contour/NLB.

```
 annotations: {
      service.beta.kubernetes.io/aws-load-balancer-ssl-cert: "arn:aws:acm:us-west-2:xxx:certificate/xxxx",
      service.beta.kubernetes.io/aws-load-balancer-cross-zone-load-balancing-enabled: "true",
      service.beta.kubernetes.io/aws-load-balancer-healthcheck-protocol: "tcp",
      service.beta.kubernetes.io/aws-load-balancer-ssl-negotiation-policy: "ELBSecurityPolicy-TLS13-1-2-2021-06",
      service.beta.kubernetes.io/aws-load-balancer-ssl-ports: "443",
      service.beta.kubernetes.io/aws-load-balancer-backend-protocol: "tcp",
      service.beta.kubernetes.io/aws-load-balancer-proxy-protocol: '*',
      service.beta.kubernetes.io/aws-load-balancer-type: "nlb"
    }
```
